### PR TITLE
feat: Introduce lightningcss-loader for webpack users

### PR DESCRIPTION
**Pull request by kdy1**
### What?

I'm recreating a PR because CI of https://github.com/vercel/next.js/pull/58712 uses `lightningcss@1.14.0` for an unknown reason.

Add an opt-in feature to use `lightningcss` instead of webpack css-loader.

### Why?

In the name of performance.

### How?

This PR is largely based on https://github.com/fz6m/lightningcss-loader by @fz6m.
(Thank you for nice work)

Closes PACK-1998
Closes PACK-2124

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e48aa015f9ea1f8e206dbb3fa999f5c9c11c0a97

`pnpm test-dev test/development/app-dir/experimental-lightningcss/experimental-lightningcss.test.ts` (PPR)
- experimental-lightningcss with default mode > in dev server > should support css modules
- experimental-lightningcss with default mode > in dev server > should support browserslist

<details>
<summary>Expand output</summary>

  ● experimental-lightningcss with default mode › in dev server › should support css modules

    expect(received).toBe(expected) // Object.is equality

    Expected: "hello world"
    Received: ""

      34 |         // Recommended for tests that check HTML. Cheerio is a HTML parser that has a jQuery like API.
      35 |         const $ = await next.render$('/')
    > 36 |         expect($('p').text()).toBe('hello world')
         |                               ^
      37 |         // We remove hash frmo the class name in test mode using env var because it is not deterministic.
      38 |         expect($('p').attr('class')).toBe('search-keyword style-module__blue')
      39 |       })

      at Object.toBe (development/app-dir/experimental-lightningcss/experimental-lightningcss.test.ts:36:31)

  ● experimental-lightningcss with default mode › in dev server › should support browserslist

    page.waitForSelector: Timeout 60000ms exceeded.
    =========================== logs ===========================
    waiting for locator('.nested')
    ============================================================

      422 |     return this.chain(() => {
      423 |       return page
    > 424 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      425 |         .then(async (el) => {
      426 |           // it seems selenium waits longer and tests rely on this behavior
      427 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:424:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| buildDuration | 11.8s | 11.7s | N/A |
| buildDurationCached | 6s | 5.1s | N/A |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +205 kB |
| nextStartRea..uration (ms) | 424ms | 429ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.4 kB | ✓ |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 29.8 kB | 29.8 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 238 B | 240 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 502 B | ✓ |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.18 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 608 B | 608 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 484 B | ✓ |
| Overall change | 484 B | 484 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 94 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 37.6 kB | 37.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.55 kB | 2.55 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/lcss-loader | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 95.7 kB | 95.7 kB | ✓ |
| app-page-tur..prod.js gzip | 96.4 kB | 96.4 kB | ✓ |
| app-page-tur..prod.js gzip | 91 kB | 91 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 90.3 kB | 90.3 kB | ✓ |
| app-route-ex...dev.js gzip | 22.2 kB | 22.2 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| app-route.ru...dev.js gzip | 21.6 kB | 21.6 kB | ✓ |
| app-route.ru..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.42 kB | 9.42 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.69 kB | 9.69 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.42 kB | 9.42 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.7 kB | 49.7 kB | ✓ |
| Overall change | 933 kB | 933 kB | ✓ |

</details>

</details>
Commit: e48aa015f9ea1f8e206dbb3fa999f5c9c11c0a97

File: packages/next/src/build/webpack/config/blocks/css/loaders/global.ts [hunk 1]
```diff
@@ -24,28 +24,42 @@ export function getGlobalCssLoader(
...
-      importLoaders: 1 + preProcessors.length,
-      // Next.js controls CSS Modules eligibility:
-      modules: false,
-      url: (url: string, resourcePath: string) =>
-        cssFileResolve(url, resourcePath, ctx.experimental.urlImports),
-      import: (url: string, _: any, resourcePath: string) =>
-        cssFileResolve(url, resourcePath, ctx.experimental.urlImports),
-    },
-  })
+  if (ctx.experimental.useLightningcss) {
```
Comment (timneutkens) [inline packages/next/src/build/webpack/config/blocks/css/loaders/global.ts:27]: We should probably reason about postcss config existing as well, especially the behavior there, for example should it run both postcss and lightningcss in a two-pass or should it run one or the other.

Comment (kdy1) [reply packages/next/src/build/webpack/config/blocks/css/loaders/global.ts:27]: Is [`296bf83` (#61327)](https://github.com/vercel/next.js/pull/61327/commits/296bf83716472f0f5f0fad85ec1e9e369d6d6735) what you want?

File: packages/next/package.json [hunk 2]
```diff
@@ -102,6 +102,7 @@
   },
   "peerDependencies": {
     "@opentelemetry/api": "^1.1.0",
+    "lightningcss": "^1.22.1",
```
Comment (kwonoj) [inline packages/next/package.json:105]: I think we want to remove dependencies to this package (https://github.com/vercel/next.js/commit/1f6feac2d4314ac14c875e47b88bf84be3cec9cd)

**Review by kwonoj — CHANGES_REQUESTED** [review]

**Comment by kwonoj** [comment]
CI still seems failing?

**Review by kwonoj — APPROVED** [review]

**Review by sokra — APPROVED** [review]

**Pull request merged**